File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/RequireFilesExistRule.java
package ai.harness.maven;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.List;

/**
 * Rule that requires specified files to exist in the project.
 */
public class RequireFilesExistRule implements HarnessCheckRule {
    private static final String CATEGORY = "requireFilesExist";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public List<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        JsonNode catNode = manifest.get(CATEGORY);
        List<String> paths = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("paths"));
        String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return paths.stream()
                .filter(path -> !HarnessCheckHelper.isSafeRegularFile(root, root.resolve(path)))
                .map(path -> new Finding(severity, CATEGORY, "missing file: " + path))
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/RequireFilesExistRule.kt
package ai.harness.gradle

import kotlinx.serialization.json.JsonObject
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isRegularFile
import kotlin.io.path.isSymbolicLink

/**
 * Rule that requires specified files to exist.
 */
class RequireFilesExistRule : HarnessCheckRule {
	override fun applies(manifest: JsonObject): Boolean {
		val category = "requireFilesExist"
		val catObj = manifest[category]?.jsonObject ?: return false
		val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
		return enabled
	}

	override fun validate(manifest: JsonObject, root: Path, psiResults: HarnessPsiResults?): Collection<Finding> {
		val category = "requireFilesExist"
		val severity = HarnessCheck.Companion.severityOf(manifest, category)
		val catObj = manifest[category]?.jsonObject ?: return emptyList()
		val parametersObj = catObj["parameters"]?.jsonObject ?: return emptyList()
		val paths = HarnessCheck.Companion.stringArrayFrom(parametersObj, "paths")
		return buildSet<Finding> {
			paths.forEach { path ->
				val p = root / path
				when {
					p.isSymbolicLink() && !HarnessCheck.Companion.isAllowedRootContractSymlink(root, p) -> add(Finding(Severity.ERROR, category, "symlink file is not allowed: $path"))
					!p.isRegularFile() -> add(Finding(severity, category, "missing file: $path"))
				}
			}
		}.toList()
	}
}
